File: node/src/wrapper.ts
import { AgentMailClient } from 'agentmail'

export class Wrapper {
    constructor(private readonly client = new AgentMailClient()) {}

    public async call(method: string, args: any) {
        const parts = method.split('.')
        const methodKey = parts.pop()

        if (!methodKey) throw new Error('Method name empty')

        let parent: any = this.client
        for (const part of parts) parent = parent[part]

        const { inbox_id, thread_id, message_id, draft_id, attachment_id, ...options } = args
        const methodArgs = [inbox_id, thread_id, message_id, draft_id, attachment_id, options].filter(Boolean)

        return await parent[methodKey].call(parent, ...methodArgs)
    }

    public async safeCall(method: string, args: any) {
        try {
            return { isError: false, result: await this.call(method, args) }
        } catch (error) {
            if (error instanceof Error) return { isError: true, result: error.message }
            else return { isError: true, result: 'Unknown error' }
        }
    }
}

File: python/src/agentmail_toolkit/wrapper.py
from typing import Optional, Any
from abc import ABC
from pydantic import BaseModel
from agentmail import AgentMail


class Wrapper(ABC):
    _client: AgentMail = None

    def __init__(self, client: Optional[AgentMail] = None):
        self._client = client or AgentMail()

    def call_method(self, method_name: str, args: dict[str, Any]) -> BaseModel:
        if hasattr(self, method_name):
            return getattr(self, method_name)(**args)
        else:
            method = self._client
            for part in method_name.split("."):
                method = getattr(method, part)

            return method(**args)
